README.md
oligo-melting
===

[![DOI](https://zenodo.org/badge/108285623.svg)](https://zenodo.org/badge/latestdoi/108285623)

A Python3 package for melting temperature calculation of oligonucleotides hybridization and secondary structures.

#### Features

* Handles DNA:DNA, RNA:RNA and DNA/RNA hybridizations.
* Corrects for both salt and chemicals.
* Compatible with UNAfold (OligoArrayAux) output for direct salt and chemicals (denaturants) correction.
* Produces melting curves for the provided sequence.
* Input either as single sequence or FASTA file.
* Slightly faster than BioPython.SeqUtils.MeltingTemp.
* Provides dG, dS and dH alongside melting temperature.

#### Limitations

* Does not handle mismatches or dangling ends
* Does not handle ambiguous bases

Installation
---

To **install**, run the following:

```
git clone http://github.com/ggirelli/oligo-melting
cd oligo-melting
sudo -H pip3 install .
```

To **uninstall** run the following from within the repository folder:

```
sudo -H pip3 uninstall oligo_melting
```

To **update**, first uninstall, and then run the following from within the repository folder.

```
git pull
sudo -H pip3 uninstall oligo_melting
sudo -H pip3 install .
```

Usage
---

### From command line

#### Duplexes

The `melt_duplex` command allows to calculate the melting temperature of a nucleic acid duplex, provided the sequence of one of the two strands.

The hybridization delta free energy calculation is based on the N-N thermodynamic values in literature and is available for DNA:DNA[3], RNA:RNA[1] and DNA:RNA[2] duplexes. The melting temperature calculation is based on Santalucia, 1998[4]. Sodium and cagnesium concentration correction is based on the work of Owczarzy et al[5][6]. Formamide correction can be performed based on two different published models[7][8].

* Use the `-t` option to specify the **type of nucleic acid duplex**.
* Use `--fa-mode` to switch between linear melting temperature **formamide-based correction**[7] and linear &Delta;G formamide-based correction[8].
* Use `--fa-mvalue` together with `--fa-mode wright` to specify the **m-value** for the formamide-based correction.
* Provide the path to a **fasta file** instead of a single sequence to calculate the melting temperature of every sequence in the file.
* Use the `-v` option to trigger the verbose mode, which provides more **details** for every single sequence.
* Use `-C` for the temperature in **degree Celsius** instead of Kelvin.
* Use `--out-curve` to specify a file where to save estimated single-sequence **melting curves** with temperature range and step around the melting temperature as defined with `--t-curve`.

#### Secondary structure

The `melt_secstr` script allows to correct the melting temperature of a nucleic acid secondary structure, previously calculated with OligoArrayAux, and to produce the corresponding melting curves.

### As a library

Import the package and use the corresponding functions.

```python3
import oligo_melting as OligoMelt

seq = "CAGTCAGTCGATC"

# Calculate melting temperature for 25uM oligos
(name, g, h, s, tm, seq) = OligoMelt.Duplex.calc_tm(seq, oligo_conc = 25e-6)
print(tm)

# Adjust for 300 mM [Na+]
tm = OligoMelt.Duplex.adj_ions(tm, 0.3, 0, seq)
print(tm)
```

The `Duplex` module contains functions for duplex hybridization and melting temperature calculation, while the `SecStr` module contains similar methods for evaluating secundary structure melting temperatures.

References
---

* [1]: Freier et al, PNAS(83), 1986;
* [2]: Sugimoto et al, Biochemistry(34), 1995.
* [3]: Allawi & Santalucia, Biochemistry(36), 1997;
* [4]: SantaLucia, PNAS(95), 1998;
* [5]: Owczarzy et al, Biochemistry(43), 2004;
* [6]: Owczarzy et al, Biochemistry(47), 2008;
* [7]: McConaughy et al, Biochemistry(8), 1969;
* [8]: Wright et al, Appl. env. microbiol.(80), 2014.

License
---

```
MIT License
Copyright (c) 2017 Gabriele Girelli
```

---

This project comes from the [potpourri](https://github.com/ggirelli/potpourri) sandbox.  \\( ﾟヮﾟ)/

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]


## [2.0.2.post1]
### Changed
- Retaining FASTA header comments in output.


## [2.0.2]
### Fixed
- Typo in `melt_duplex`.
- Duplicated output from `melt_duplex` when `-F` is not used.
- Wrong variable name in `duplex` module.

### Changed
- Version tag now stored in const module.
- Now using `black` for code formatting.



##[2.0.1.post5]
### Added
- Version control of scripts and `--version` flag.
- `melt duplex` output in FASTA format.
- FASTA header key-value delimiter as option.

### Changed
- FASTA input parsed with biopython SimpleFastaParser.
- Using `poetry` for dependency management.
- Entry points changed to base keyword `melt`.

### Fixed
- Proper script help page formatting.
- Bug that duplicated every line in the output.

## [2.0.1]
### Changed
- Re-packed software for pypi.

### Added
- Default parameter values.

## [1.4.5]
### Fixed
- Fixed issue with new interaction type labeling system.

### Added
- Automatic R packages installation (argparser and ggplot2).

### Changed
- Separated curve calculation and fraction calculation to different functions.
- Harmonized `melt_X.py` scripts output.
- Better fasta reading management: not kept in memory and considering sequences with same header as separate fasta items.

### Added
- `plot_melt_curves_coupled.R` for single-oligo coupled melting curve plot.
- Allowed for initial formamide and sodium concentration for melting temperature correction.

## [1.4.4]
### Changed
- Renamed `oligomeltlib.py` to `meltlib.py`.
- Renamed `oligomelt.py` to `melt_duplex.py`.

### Added
- `melt_second.py` to perform formamide correction and melting curve calculation for secondary structures predicted with OligoArrayAux.

## [1.4.3]
### Added
- Melting curve plotting script (from fish-conditions repo).

### Changed
- Moved functions to separate library

### Fixed
- Now melting curves show the proper (inverted) dissociation curve.

## [1.4.2]
### Changed
- Now using user-defined formamide m-value for wright correction.

## [1.4.0]
### Added
- New feature: formamide correction.

## [1.3.0]
### Added
- New feature: temperature curve calculation. Proper fasta input.

## [1.2.2]
### Fixed
- Allawi and Freier thermodynamic tables.

## [1.2.1]
### Fixed
- Introduced Sugimoto (DNA:RNA) thermodynamic table.

## [1.2.0]
### Added
- New feature: DNA/RNA and RNA/DNA duplex calculation.

## [1.1.0]
### Added
- New input file mode.

### Fixed
- Mg2+ correction now skips Na+ correction.

## [1.0.0]

* [unreleased] https://github.com/ggirelli/oligo-melting
* [2.0.2] https://github.com/ggirelli/oligo-melting/releases/tag/v2.0.2
* [2.0.1] https://github.com/ggirelli/oligo-melting/releases/tag/v2.0.1
* [2.0.0] https://github.com/ggirelli/oligo-melting/releases/tag/v2.0.0
* [1.4.5] https://github.com/ggirelli/oligo-melting/releases/tag/v1.4.5
* [1.4.4] https://github.com/ggirelli/oligo-melting/releases/tag/v1.4.4
* [1.4.3] https://github.com/ggirelli/oligo-melting/releases/tag/v1.4.3
* [1.4.2] https://github.com/ggirelli/oligo-melting/releases/tag/v1.4.2
* [1.4.0] https://github.com/ggirelli/oligo-melting
* [1.3.0] https://github.com/ggirelli/oligo-melting
* [1.2.2] https://github.com/ggirelli/oligo-melting
* [1.2.1] https://github.com/ggirelli/oligo-melting
* [1.2.0] https://github.com/ggirelli/oligo-melting
* [1.1.0] https://github.com/ggirelli/oligo-melting
* [1.0.0] https://github.com/ggirelli/oligo-melting


oligomelt.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# ------------------------------------------------------------------------------
# 
# Author: Gabriele Girelli
# Email: gigi.ga90@gmail.com
# Version: 1.4.3
# Date: 20170711
# Project: oligo characterization
# Description:  calculate melting temperature of a provide NA duplex
# 
# Changelog:
#  1.4.3: fixed melting curves (inverted).
#  1.4.2: changed to user-defined formamide m-value for wright correction.
#  1.4.0: added formamide correction.
#  1.3.0: added temperature curve calculation. Proper fasta input.
#  1.2.2: fixed allawi and freier tables.
#  1.2.1: fixed sugimotod table.
#  1.2.0: DNA/RNA and RNA/DNA duplex calculation.
#  1.1.0: input file mode. Fixed Mg2+ correction.
#  1.0.0: first implementation.
# 
# References:
#  [1] Freier et al, PNAS(83), 1986;
#  [2] Sugimoto et al, Biochemistry(34), 1995.
#  [3] Allawi & Santalucia, Biochemistry(36), 1997;
#  [4] SantaLucia, PNAS(95), 1998;
#  [5] Owczarzy et al, Biochemistry(43), 2004;
#  [6] Owczarzy et al, Biochemistry(47), 2008;
#  [7] McConaughy et al, Biochemistry(8), 1969;
#  [8] Wright et al, Appl. env. microbiol.(80), 2014.
# 
# ------------------------------------------------------------------------------



# DEPENDENCIES =================================================================

import argparse
import os
import re
import sys

from lib.oligomeltlib import *

# PARAMETERS ===================================================================

# Add script description
parser = argparse.ArgumentParser(
    description = '''
Calculate melting temeprature of a DNA duplex at provided [oligo],
[Na+], [Mg2+]. Either provide an oligo sequence or a file with one oligo
per line (and use -F option). References:
 [1] Freier et al, PNAS(83), 1986;
 [2] Sugimoto et al, Biochemistry(34), 1995.
 [3] Allawi & Santalucia, Biochemistry(36), 1997;
 [4] SantaLucia, PNAS(95), 1998;
 [5] Owczarzy et al, Biochemistry(43), 2004;
 [6] Owczarzy et al, Biochemistry(47), 2008;
 [7] McConaughy et al, Biochemistry(8), 1969;
 [8] Wright et al, Appl. env. microbiol.(80), 2014.
''')

# Add mandatory arguments
parser.add_argument('seq', type = str, nargs = 1, help = '''
    DNA duplex sequence (one strand only) or path to a FASTA file (use with -F).
    ''')

# Add arguments with default value
parser.add_argument('-t', '--type', type = str, nargs = 1,
    help = '''Duplex type. Possible values: DNA:DNA (based on ref.3, default),
    RNA:RNA (based on ref.1), DNA:RNA (based on ref.2., given DNA sequence)
    or RNA:DNA (based on ref.2, given RNA sequence). The first nucleic acid type
    indicates the provided sequence.''',
    choices = ['DNA:DNA', 'RNA:RNA', 'RNA:DNA', 'DNA:RNA'],
    default = ['DNA:DNA'])
parser.add_argument('-o', '--oconc', metavar = "oligo_conc",
    type = float, nargs = 1,
    help = '''Oligonucleotide concentration [M].
    Default: 0.25e-6 M''',
    default = [0.25e-6])
parser.add_argument('-n', '--naconc', metavar = "na_conc",
    type = float, nargs = 1,
    help = '''Na+ concentration [M].
    Default: 50e-3 M''',
    default = [50e-3])
parser.add_argument('-m', '--mgconc', metavar = "mg_conc",
    type = float, nargs = 1,
    help = '''Mg2+ concentration [M]. Note: Mg2+ correction overwrites Na+
    correction. Default: 0 M''',
    default = [0])
parser.add_argument('-f', '--faconc', type = float, nargs = 1,
    metavar = 'fa_conc', help = '''Formamide concentration in %%(v,v).''',
    default = [0])
parser.add_argument('--fa-mode', type = str, nargs = 1,
    metavar = 'fa_mode', help = '''Mode of formamide correction. "mcconaughy"
    for classical -0.72%%FA correction from ref. 7, "wright" for single reaction
    model correction from ref.8 (default).''',
    choices = ['mcconaughy', 'wright'], default = ['wright'])
parser.add_argument('--fa-mvalue', type = str, nargs = 1, metavar = 'm',
    help = '''Specify the formamide m-value to be used with the wright
    correction model. Use either a single value "x" or two values "xL+y" where
    L is the probe length. Default: 0.1734''', default = ["0.1734"])
parser.add_argument('--t-curve', type = float, nargs = 2,
    metavar = ('range', 'step'), help = '''Temperature range and step for
    melting curve generation. Use --make-curve option to generate the curve.
    Default: 20 degC range and 0.5 degC step.''', default = [20, 0.5])
parser.add_argument('--out-curve', type = str, nargs = 1, metavar = "outname",
    help = '''Path to output table containing tabulated curves.''',
    default = [False])

# Add flags
parser.add_argument('-C', '--celsius',
    dest = 'celsius', action = 'store_const',
    const = True, default = False,
    help = 'Output temperature in Celsius degrees. Default: Kelvin')
parser.add_argument('-F', '--usefile',
    dest = 'usefile', action = 'store_const',
    const = True, default = False,
    help = 'Use when a file path is provided instead of a single sequence.')
parser.add_argument('-v', '--verbose',
    dest = 'verbose', action = 'store_const',
    const = True, default = False,
    help = 'Verbose output.')

# Parse arguments
args = parser.parse_args()

# Assign to in-script variables ------------------------------------------------

# Input oligo
seq = args.seq[0]

# Concentrations
oligo_conc = args.oconc[0]
na_conc = args.naconc[0]
mg_conc = args.mgconc[0]

# Thermodynamic table
tt_mode = args.type[0]

# Temperature units
celsius = args.celsius

# File as input
use_file = args.usefile

# Verbose mode
is_verbose = args.verbose

# Melting curve
curve_range = args.t_curve[0]
curve_step = args.t_curve[1]
curve_outpath = args.out_curve[0]
do_curve = False != curve_outpath

# Formamide
fa_conc = args.faconc[0]
fa_mode = args.fa_mode[0]
fa_mval_s = args.fa_mvalue[0]

# Additional checks ------------------------------------------------------------

# Check proper curve step/range pair
if curve_step > curve_range / 2:
    sys.exit("!!!ERROR! Curve step must be smaller than curve range.")
curve_range -= curve_range % curve_step

# Remove output curve file if it exists
if do_curve and os.path.isfile(curve_outpath):
    os.remove(curve_outpath)

# Evaluate formamide m-value
parsed_mval = False
mregex = ["(([+-]?[0-9\.]*)L([+-][0-9\.]*))", "([+-]?[0-9\.]*)"]

# Search for xL+y m-value format
msearch = re.search(mregex[0], fa_mval_s)
if not type(None) is type(msearch):
    mgroups = msearch.groups()
    if fa_mval_s == mgroups[0]:
        mvalue = lambda x: float(mgroups[1]) * x + float(mgroups[2])
        parsed_mval = True

# Search for x m-value format
msearch = re.search(mregex[1], fa_mval_s)
if not parsed_mval and not type(None) is type(msearch):
    mgroup = msearch.groups()[0]
    if fa_mval_s == mgroup:
        mvalue = lambda x: float(mgroup)
        parsed_mval = True

# Trigger error otherwise
if not parsed_mval:
    sys.exit("!!!ERROR! Unexpected formamide m-value format Check help page.")

# Fix m-value label
if not fa_mode in ["wright"]:
    fa_mval_s = "-"

# Thermodynamic tables
tt = {
    'RNA:RNA'   : freier,
    'DNA:DNA'   : allawi,
    'RNA:DNA'   : sugimotor,
    'DNA:RNA'   : sugimotod
}

# FUNCTIONS ====================================================================

# RUN ==========================================================================

# Build argument dictionary
data = {
    'oligo_conc' : oligo_conc,
    'na_conc' : na_conc,
    'mg_conc' : mg_conc,
    'fa_conc' : fa_conc,
    'fa_mode' : fa_mode,
    'fa_mval_s' : fa_mval_s,
    'mvalue' : mvalue,
    'tt' : tt,
    'tt_mode' : tt_mode,
    'celsius' : celsius,
    'is_verbose' : is_verbose,
    'do_curve' : do_curve,
    'curve_step' : curve_step,
    'curve_range' : curve_range,
    'curve_outpath' : curve_outpath
}

# CALCULATE --------------------------------------------------------------------

if not use_file:
    # Single sequence case
    data['name'] = 'seq'
    data['seq'] = seq
    melt_calc(**data)
else:
    # Empty sequence dictionary
    fastad = {}

    # Input file case
    curr_head = ""
    with open(seq) as fin:
        for row in fin:
            if ">" == row[0]:
                curr_head = row[1:].strip()
            else:
                if curr_head in fastad.keys():
                    fastad[curr_head] = fastad[curr_head] + row.strip()
                else:
                    fastad[curr_head] = row.strip()

    # Calculate for each fasta item
    for (name, seq) in fastad.items():
        data['name'] = name
        data['seq'] = seq
        melt_calc(**data)

# END ==========================================================================

################################################################################
